geoAssembler/optimiser/utility.py

```diff
@@ -0,0 +1,131 @@
+import numpy as np
+
+from extra_data import stack_detector_data
+from pyFAI.detectors import Detector
+from pyFAI.azimuthalIntegrator import AzimuthalIntegrator
+
+from extra_data.reader import DataCollection
+from extra_geom.detectors import DetectorGeometryBase
+from typing import Union, Tuple
+
+
+class Integrator:
+    """
+    Object wrapping pyFAI and extra-geom geometries to probide a more
+    convenient way of integrating 2d detector images.
+
+    Creating the integrator requires a geometry, as well as the distance
+    from the sample to the detector in mm. If unknown, 200mm seems to be
+    a good 'normal' guess for AGIPD.
+
+    Once created, use the `integrate2d` method to get a 2d unrolled view
+    of the detector image. You can them sum over the image to get a 1d
+    integration result.
+    """
+
+    def __init__(self, geom: DetectorGeometryBase,
+                 sample_dist_mm: Union[int, float], unit: str = "2th_deg"):
+        self.unit = unit
+        self.sample_dist_mm = sample_dist_mm
+
+        fakedata = np.zeros(geom.expected_data_shape)
+        fakeimage, centre_geom = geom.position_modules_fast(fakedata)
+        self.size = fakeimage.shape
+
+        #  pyFAI takes in coordinates in the opposite order to those
+        #  produced by extra-geom, so swap it around here
+        self.centre = [centre_geom[1], centre_geom[0]]
+
+        self.detector = Detector(
+            pixel1=geom.pixel_size,
+            pixel2=geom.pixel_size
+        )
+
+        ai = AzimuthalIntegrator(detector=self.detector)
+        ai.setFit2D(self.sample_dist_mm, self.centre[0], self.centre[1])
+        self.ai = ai
+
+        self.radius = ((self.size[0]/2)**2 + (self.size[1]/2)**2)**(1/2)
+        self.azimuth_bins = self.radius * (self.size[0]/self.size[1])
+
+    def integrate2d(self, frame: np.ndarray, centre_offset=None):
+        """
+        Unroll the image - changes the axis from cartesian x/y to
+        polar radius/azimuthal angle.
+
+        A sum over the angle can be performed to get a result for
+        the azimuthal integration.
+
+        Returns a pyFAI `Integrate2dResult` object, check pyFAI
+        docs for more information.
+        """
+        if centre_offset is not None:
+            self.ai.setFit2D(
+                self.sample_dist_mm,
+                centre_offset[0] + self.centre[0],
+                centre_offset[1] + self.centre[1]
+            )
```

Comment:
Made it stateless, and changed from set_FIT2D to set_pyFAI, however this has (again) confused me about which index is x and y in the centre and in the offset.

Moving from `set_FIT2D` to `set_pyFAI` meant I had to change the order of the centre and centre offsets, now I'm a bit confused about the correct order of x and y, but the tests pass and the notebook produces the same results even though there's probably a mistake somewhere in the code.

Think two of my mistakes cancel out, I'll look into this a bit more later :wink:

Reply:
Yep, coordinate order is a regular source of headaches. It's all because images tend to be stored row by row, so you index pixels (y, x) in languages that follow C rather than Fortran.